### PR TITLE
fix race condition when scanning short ranges

### PR DESCRIPTION
**Pull request by NikitaKovaljov**
- This change fix race condition in ARP while scanning hosts 

## Reference to Issue

-  #16616

## Verification without changes

- run ```./msfconsole```

- use ```ipv6_neighbor.rb``` module without changes. 

- set rhosts ```10.0.0.1-6```

- run module ```run or exploit```

### Current behavior
```
[*] Discovering IPv4 nodes via ARP...
[+]           10.0.0.2 ALIVE
[*] Discovering IPv6 addresses for IPv4 nodes...
[*] 
[*] Scanned 6 of 6 hosts (100% complete)
[*] Auxiliary module execution completed
```
## Verification with changes

- run ```./msfconsole```

- use ```ipv6_neighbor.rb``` module with changes. 

- set rhosts ```10.0.0.1-6```

- run module ```run or exploit```

### Fixed behavior
```
[*] Discovering IPv4 nodes via ARP...
[+]           10.0.0.2 ALIVE
[+]           10.0.0.6 ALIVE
[*] Discovering IPv6 addresses for IPv4 nodes...
[*] 
[*] Scanned 6 of 6 hosts (100% complete)
[*] Auxiliary module execution completed
```

### CODE REVIEW

File: modules/auxiliary/scanner/discovery/ipv6_neighbor.rb [hunk 1]
```diff
@@ -74,7 +74,7 @@ def run_batch(hosts)
         end
       end
 
-      etime = ::Time.now.to_f + (hosts.length * 0.05)
+      etime = ::Time.now.to_f + 1
```
Comment (smcintyre-r7) [inline modules/auxiliary/scanner/discovery/ipv6_neighbor.rb:77]: Thinking we should perhaps keep the time scaling but with a minimum value of one like you're using here. That would ensure we're always waiting at least 1 second.

```suggestion
      etime = ::Time.now.to_f + [1, hosts.length * 0.05].max
```

**Comment by smcintyre-r7** [comment]
Yeah that makes sense. A configurable sleep probably would be the best option. I wasn't entirely sure that the time should scale with the hosts either, in which case I was thinking a minimum was probably the safest option.

Right now it looks like the `TIMEOUT` datastore option is defaulting to 500, probably from the [Capture](https://github.com/rapid7/metasploit-framework/blob/e6e181618ab8026716527cc59dcc5c8520a25ec2/lib/msf/core/exploit/capture.rb#L32) mixin that's included. If you want to use that datastore option, which again makes sense, you probably just want to update that [default value](https://github.com/rapid7/metasploit-framework/blob/e6e181618ab8026716527cc59dcc5c8520a25ec2/modules/exploits/linux/local/cron_persistence.rb#L46).

**Comment by smcintyre-r7** [comment]
@NikitaKovaljov just wondering if you have any thoughts regarding the above comments or if we want to keep things as is?

Just want to make sure this isn't forgotten about.

**Comment by NikitaKovaljov** [comment]
> @NikitaKovaljov just wondering if you have any thoughts regarding the above comments or if we want to keep things as is?
> 
> Just want to make sure this isn't forgotten about.

Hi, 

Just now added our idea with datastore['TIMEOUT'] - tested locally seems legit. 

Now waiting for github check to complete.

**Review by smcintyre-r7 — APPROVED** [review]
Changes look good now and I confirmed this fixed the original issue. In the screenshot below you can see the options are the same and when scanning 2 hosts, the race condition would cause the unpatched version to miss the hosts that were online. I ran this a few times and observed that the patched module consistently reported the results while the unpatched was hit or miss.

image


I'll get this landed in a moment, thanks for the bug fix @NikitaKovaljov !

**Pull request merged**